### ISSUE TITLE
[Question]: Is playwhright-go thread-safe if multiple instances are created in separate go routines?

### ISSUE THREAD

**Issue by martingrzzler** [playwright-community/playwright-go#492]
I'm using this awesome library for web scraping but when running browsers in parallel I get random crashes. 
So this might be because the library is not thread-safe. In the following example I create two different instances of Chrome with `playwhright.Run`. Is this save to do?

```go
package main

import (
	"sync"

	"github.com/crawler/internal/crawl"
)

func main() {

	wg := sync.WaitGroup{}

	wg.Add(1)
	go func() {
		defer wg.Done()
                // creates a new instances with playwhright.Run
		b1, err := crawl.LauchBrowser(crawl.LaunchOptions{
			Headless: false,
		})
		if err != nil {
			panic("failed to launch browser" + err.Error())
		}
		for i := 0; i < 10; i++ {
			p, err := crawl.NewPage(b1)
			if err != nil {
				panic("failed to create new page" + err.Error())
			}

			_, err = crawl.ScrapeHomePage(p, "https://google.com")
			if err != nil {
				panic("failed to scrape home page" + err.Error())
			}

			err = p.Context().Close()
			if err != nil {
				panic("failed to close context" + err.Error())
			}
		}

	}()

	wg.Add(1)
	go func() {
		b2, err := crawl.LauchBrowser(crawl.LaunchOptions{
			Headless: false,
		})
		if err != nil {
			panic("failed to launch browser" + err.Error())
		}
		defer wg.Done()

		for i := 0; i < 10; i++ {
			p, err := crawl.NewPage(b2)
			if err != nil {
				panic("failed to create new page" + err.Error())
			}

			_, err = crawl.ScrapeHomePage(p, "https://example.com")
			if err != nil {
				panic("failed to scrape home page" + err.Error())
			}

			err = p.Context().Close()
			if err != nil {
				panic("failed to close context" + err.Error())
			}

		}
	}()

	wg.Wait()
}


```

**Comment by canstand** [comment]
What you should do is create a new `BrowserContext` per goroutine, instead of launching multiple browsers.

**Comment by martingrzzler** [comment]
@canstand So using multiple `BrowserContext` in each goroutine is safe then? 
I'm asking because of this comment: https://github.com/playwright-community/playwright-go/issues/119#issuecomment-892615173

**Comment by canstand** [comment]
It can be considered safe for your use case. BrowserContext is isolated from each other, which is equivalent to opening multiple browser windows and each has its own temporary data directory and profile. 

You need to be careful to avoid deadlocks like #481 and #391 .

**Comment by mxschmitt** [comment]
Thanks for the report, and apologies for the long delay here.

To answer the question directly: **yes, playwright-go is safe to use concurrently**, with one design note and one bug that was just fixed.

### Recommended pattern
Don't launch a separate browser (driver process) per goroutine. Launch **one browser** and create **one `BrowserContext` per goroutine**. Contexts are fully isolated from each other (separate storage, cookies, profile) — equivalent to separate browser windows — so each goroutine can drive its own context/pages safely in parallel.

### The "random crashes" were a real deadlock — now fixed
The hangs you were seeing were almost certainly the blocking-call-in-event-handler deadlock (#391, #481, #398, #574). Event handlers (`OnResponse`, `OnRequest`, `OnDOMContentLoaded`, …) ran synchronously on the single connection dispatch goroutine — the same goroutine that delivers protocol replies. So any blocking server call from inside a handler (`Response.Body()`, `HeadersArray()`, `Page.Evaluate()`, …) waited forever for a reply only that goroutine could deliver.

This is fixed in #603 (merged to `main`): when a blocking call is made on the dispatch goroutine, the receive loop is pumped re-entrantly until the reply arrives, instead of blocking.

### Verified before/after
Using the regression test added in #603 (`Response.HeadersArray()` + `Body()` inside `OnResponse`):

- **Before the fix** (parent commit) — deadlocks; the dispatcher hangs and the test never returns:
  ```
  connection.Dispatch → channel.Emit → callHandlers → [OnResponse handler]
    → responseImpl.HeadersArray → protocolCallback.waitResult   (blocked forever)
  ```
- **After the fix** (`main`):
  ```
  --- PASS: TestBlockingServerCallInsideEventHandler (0.08s)
  ```

Closing as resolved — it'll be in the next release. If you still hit crashes with the one-browser + context-per-goroutine pattern on the latest version, please open a new issue with a runnable repro. Thanks!
